### PR TITLE
got: update to 0.92

### PR DESCRIPTION
**Pull request by artkiver**
###### Type(s)

- [x] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on

macOS 14.0 23A5328b arm64
Xcode 15.0 15A5219j

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [x] checked your Portfile with `port lint --nitpick`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?
- [x] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

### CODE REVIEW

File: devel/got/Portfile [hunk 1]
```diff
@@ -20,15 +20,18 @@ long_description    Game of Trees is a version control system which \
...
+                    size 1265096
 depends_lib         path:lib/libssl.dylib:openssl \
                     port:ncurses \
                     port:ossp-uuid \
                     port:libevent
 depends_build       port:bison \
                     port:pkgconfig
+variant gotd description {build with got's server daemon implementation} {
+configure.args --enable-gotd
+}
```
[the remark this answers is not in the document]
Comment (artkiver) [reply devel/got/Portfile:34]: > why not enable this unconditionally and use a variant instead? There are no extra dependencies needed so what's the drawback of always enabling this option?

@reneeotten  That's a good question and certainly a possibility.

Previously, gotd was compiled by default. I added the variant more to reflect the nuance with this change in the release of the -portable branch which broke it out into a configure option per the notes here: http://gameoftrees.org/releases/portable/CHANGELOG.

Excerpt:

```
This release also introduces 'gotd', which is got's server implementation.  By
extension, other tools are also introduced to work with gotd, and they are:

* gitwrapper
* gotctl
* gotsh

For more information about gotd, see gotd(8), as well as the above command man
pages where necessary.  Note that gotd's portable implementation uses chroot()
to ensure some separation; this still requires the main gotd binary to run as
the root user.

To enable gotd and the corresponding other tools which work with it, you must
pass the '--enable-gotd' ./configure flag.
```

Also keeping in mind Thomas Adam's notes here: https://bsd.network/@thomasadam/110974000446820792. 

Specifically,

```
Please note:  the one area I have not been able to look into with supporting gotd in -portable, is MacOS.

It compiles just fine, and chroot() is something MacOS understands, so it should Just Work(tm), but if there are any issues, drop me a note!
```

Essentially, the main change as I understand it and recall (and from some mailing list discourse) was to attempt to add some rudimentary security to gotd via chroot, in lieu of macOS and other OSes not having pledge/unveil type mechanisms such as OpenBSD. Additionally, since gotd is a daemon and requires root access unlike the rest of Got, it seemed as if it might be worth making it into a variant rather than a default?

Admittedly, the thought came to me later. This PR having two commits (which I wasn't able to squash, but that's on me) reflects that.

By default, gotd won't be built without that configure parameter, so we could leave it as my initial commit and remove the variant entirely relying on that for users advanced enough to read the source and build it accordingly? I thought offering a variant might be a more versatile way and save more advanced users a little bit of effort, but I'm open to alternative recommendations!

I had also contemplated adding a notes field to articulate the differences with this variant.

Another consideration I had was to make this variant a default, but I tend to side within the realms of keeping code secure by default and thus anything dependent upon root access such as gotd is something I to which I at least give pause. In this instance, I thought about whether it's worthwhile to build gotd for all users when hopefully only more advanced users will be requiring it and will have sense of mind to look a little more deeply for what they need as it may require more thoughtful configuration.

Hopefully that at least provides my rationale, though it need not be the be all end all. I appreciate you taking a look and inquiring.

[the remark this answers is not in the document]
Comment (ThomasAdam) [reply devel/got/Portfile:34]: Hi,

I think there's some confusion here.

@artkiver, regarding "Previously, gotd was compiled by default.", this is not true.  In -portable, gotd (and associated tools), have always been behind the `--enable-gotd` configure option -- long before it was merged and released in 0.92

My suggestion is to always pass `--enable-gotd`.  There's no issue in doing so as there's no extra dependencies.  Furthermore, until the end user configures gotd nothing will happen.  It's harmless to have these applications installed out-of-the-box, IMO.

As for the release note about gotd(8) not being tested on MacOS -- functionally it will work, but there's perhaps going to be a need from a packaging perspective to include some of the extra configure flags, such as `--with-gitwrapper-git-libexec-path` and `--with-gitwrapper-git-libexec-path`.

[the remark this answers is not in the document]
Comment (ryandesign) [reply devel/got/Portfile:34]: I agree with always enabling gotd, and removing the variant.

Regarding "perhaps" needing those extra configure args, how would we know if we need them?

[the remark this answers is not in the document]
Comment (artkiver) [reply devel/got/Portfile:34]: Consensus seems to be to remove the variant and make --enable-gotd the default, which is OK by me! Thank you @ThomasAdam for sharing your perspective on this despite it not being your normal m.o. with regards to package maintenance!

Thank you too for chiming in @ryandesign you are articulating some questions lingering in my mind as well.

Namely, regarding --with-gitwrapper-git-libexec-path and --with-gotd-empty-path I would appreciate more guidance or recommendations with what appropriate paths for MacPorts might be (assuming those configure flags take parameters of a path given their names?). Certainly, adding those to configure.args and rebuilding doesn't seem to throw any errors without path declarations, but I am guessing it's not exactly how it's intended to be built? (I'll update this PR with a push showing another Portfile variant with what I have been testing to attempt to illustrate my point here momentarily)

TL;DR, additional input is welcome by me here, not to harp on this but since @ThomasAdam wrote elsewhere, "It compiles just fine, and chroot() is something MacOS understands, so it should Just Work(tm), but if there are any issues, drop me a note!" I'll take advantage of the offer to assist! ;)

You are at least correct about it compiles just fine though! Phew!

File: devel/got/Portfile [hunk 2]
```diff
@@ -20,15 +20,16 @@ long_description    Game of Trees is a version control system which \
...
+checksums           rmd160 d4bf70c788b6c8b66735f030bddd58e6a4eafc50 \
+                    sha256 1da342c606e5d1ec2f23dea3afdfc1809a61aa8402e4fd9ab63e5ae756e3f7d7 \
+                    size 1265096
 depends_lib         path:lib/libssl.dylib:openssl \
                     port:ncurses \
                     port:ossp-uuid \
                     port:libevent
 depends_build       port:bison \
                     port:pkgconfig
+configure.args --enable-gotd --with-gitwrapper-git-libexec-path --with-gotd-empty-path
```
Comment (ryandesign) [inline devel/got/Portfile:32]: Can this be changed to the usual indentation?
```suggestion
configure.args      --enable-gotd \
                    --with-gitwrapper-git-libexec-path \
                    --with-gotd-empty-path
```

File: devel/got/Portfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (ThomasAdam) [reply devel/got/Portfile]: Well, the git-wrapper installation instructions are in `README.portable.  The `—with-gitwrapper*` flag is needed to tell gitwrapper where it can find the installed git libexec helpers.  As for `—with-good-empty-path`, this is needed so good can chroot into an empty directory. This defaults to `/var/empty` which exists on BSD systems. Not sure about MacOS, but if it does exist, you won’t need to use this flag. For Linux and other systems, we might suggest `/var/run/gotd`.  This will need to exist though.

[the remark this answers is not in the document]
Comment (artkiver) [reply devel/got/Portfile]: > Well, the git-wrapper installation instructions are in `README.portable`. The `—with-gitwrapper*` flag is needed to tell gitwrapper where it can find the installed git libexec helpers. As for `—with-good-empty-path`, this is needed so good can chroot into an empty directory. This defaults to `/var/empty` which exists on BSD systems. Not sure about MacOS, but if it does exist, you won’t need to use this flag. For Linux and other systems, we might suggest `/var/run/gotd`. This will need to exist though.

Thank you @ThomasAdam  for the additional insights!

As far as I can tell (at least locally) macOS does have /var/empty and I am pretty sure I have seen some other MacPorts use that as well, so I guess we can probably omit the -with-gotd-empty-path configure argument?

Thanks for the pointer to README.portable too! D'oh. In this instance, macOS ships with git in /usr/bin/git but MacPorts users may also have a version of git installed via MacPorts which would presumably be /opt/local/bin/git so I am a little less certain how best to handle that off the top of my head (I may finally be feeling the effects of too little sleep).

For the sake of others participating in this conversation, I am going to copy/paste the salient excerpt from README.portable below and probably sleep on this before going further and again, welcome suggestions from other MacPorts contributors on what makes good sense.


```
A gotd server can be used without gitwrapper in the following cases:

1) The Git client's user account has gotsh configured as its login shell.

2) The Git client's user account sees gotsh installed under the names
   git-receive-pack and git-upload-pack, and these appear in $PATH before
   the corresponding Git binaries if Git is also installed. Setting up the
   user's $PATH in this way can require the use of SetEnv in sshd_config.

The above cases can be too restrictive. For example, users who have regular
shell access to the system may expect to be able to serve Git repositories
from their home directories while also accessing repositories served by gotd.

Once gitwrapper has been installed correctly it provides an out-of-the box
experience where both gotd and Git "just work".
However, this will require coordination with the system's Git installation
and/or distribution package because the names of two specific Git programs
will be overlapping: git-upload-pack and git-receive-pack

If the gitwrapper tool will be used then it must replace git-receive-pack
and git-upload-pack in /usr/bin. This is usually achieved by replacing the
regular Git binaries in /usr/bin with symlinks to gitwrapper:

```
-rwxr-xr-x 1 root root 1019928 Aug 24 00:16 /usr/bin/gitwrapper
lrwxrwxrwx 1 root root 10 Aug 20 12:40 /usr/bin/git-receive-pack -> gitwrapper
lrwxrwxrwx 1 root root 10 Aug 20 12:40 /usr/bin/git-upload-pack -> gitwrapper
```

The Git binaries remain available in Git's libexec directory, which is set
when Git gets compiled. On Debian it defaults to /usr/lib/git-core.
This same path must be given to Got's configure script at build time to
allow gitwrapper to find Git's binaries:

```
  ./configure --with-gitwrapper-git-libexec-path=/usr/lib/git-core
```

Once gitwrapper is found in /usr/bin under the names git-receive-pack and
git-upload-pack, any Git repositories listed in /etc/gotd.conf will be
automatically served by gotd, and any Git repositories not listed in
/etc/gotd.conf will be automatically served by regular Git's git-upload-pack
and git-receive-pack. The client's login shell or $PATH no longer matter,
and a peaceful co-existence of gotd and Git is possible.

We recommend that distribution packagers take appropriate steps to package
gitwrapper as a required dependency of gotd. It is also possible to install
gitwrapper without installing gotd. As long as /etc/gotd.conf does not exist
or no repositories are listed in /etc/gotd.conf there will be no visible
change in run-time behaviour for Git users since gitwrapper will simply run
the standard Git tools.
In the OpenBSD ports tree both the regular git package and the gotd package
are depending on gitwrapper, and the git package no longer installs the
git-receive-pack and git-upload-pack programs in /usr/local/bin.
```

[the remark this answers is not in the document]
Comment (ryandesign) [reply devel/got/Portfile]: > ```
> If the gitwrapper tool will be used then it must replace git-receive-pack
> and git-upload-pack in /usr/bin. This is usually achieved by replacing the
> regular Git binaries in /usr/bin with symlinks to gitwrapper:
> 
> -rwxr-xr-x 1 root root 1019928 Aug 24 00:16 /usr/bin/gitwrapper
> lrwxrwxrwx 1 root root 10 Aug 20 12:40 /usr/bin/git-receive-pack -> gitwrapper
> lrwxrwxrwx 1 root root 10 Aug 20 12:40 /usr/bin/git-upload-pack -> gitwrapper
> ```

The contents of /usr/bin are protected by System Integrity Protection; nothing other than Apple can change what's there. And the programs installed by the git port are owned by the git port; it would not be proper for any other port to attempt to change them. So I don't think this gitwrapper thing is happening.

[the remark this answers is not in the document]
Comment (ryandesign) [reply devel/got/Portfile]: > The `—with-gitwrapper*` flag is needed to tell gitwrapper where it can find the installed git libexec helpers. As for `—with-good-empty-path`, this is needed so good can chroot into an empty directory.

It seems that `--with-gitwrapper-git-libexec-path` is interpreted as `--with-gitwrapper-git-libexec-path=yes` and `--with-gotd-empty-path` is interpreted as `--with-gotd-empty-path=yes`. It is using the path `yes` for both options. That is of course not a valid path and is not going to work for anything so it's surprising that the configure script didn't complain about this invalid input and exit with an error.

[the remark this answers is not in the document]
Comment (artkiver) [reply devel/got/Portfile]: Thanks for looking into this more deeply @ryandesign  . Based on that evaluation, it seems as if it's probably better to omit those configure flags from the Portfile for now, perhaps I should add a notes field that configuring gotd is left as an exercise to the reader?

e.g. something such as:

```
notes "configuration of gotd is a more involved than this MacPort currently \
addresses. At a minimum, it is recommended to reinstall with \
--with-gitwrapper-git-libexec-path added to the configure.args of the MacPorts\
Portfile with a path specific to your environment. \
Consulting the gotd.conf, gotd, gotctl and gotsh\
man pages is also recommended for users wishing to take advantage of gotd's
additional functionality.
"
```
I must admit, a part of me is tempted to just omit all the configure.args including --enable-gotd and leave that level of usage for those who are willing and able to work with source directly.

[the remark this answers is not in the document]
Comment (ryandesign) [reply devel/got/Portfile]: If gitwrapper is essential to the operation of gotd, then I agree gotd should not be enabled, and I would leave out the notes; they will just confuse any users who don't need gotd, which is presumably all existing users of the port.

[the remark this answers is not in the document]
Comment (artkiver) [reply devel/got/Portfile]: > If gitwrapper is essential to the operation of gotd, then I agree gotd should not be enabled, and I would leave out the notes; they will just confuse any users who don't need gotd, which is presumably all existing users of the port.

I am OK with taking that approach for the time being. In that instance, is there a way through the GitHub UI to discard the subsequent commits and just retain the first one before I added the variant?

[the remark this answers is not in the document]
Comment (artkiver) [reply devel/got/Portfile]: In lieu of seeing anything useful in the GitHub UI, I tried this locally:

```
git revert b280d80f2c389f52457cd9bbcedb098a5da46eec
Auto-merging devel/got/Portfile
[got0.92_uno b43b0afdbe7] Revert "got: update to 0.92"
 1 file changed, 4 insertions(+), 4 deletions(-)
```
Though, the WiFi at the library where I presently am blocks SSH so I can't actually push that change. Additionally, I think I may have chosen the wrong commit ID (the initial one, rather than subsequent ones) so I am not sure if I should push that change? Currently seeing if there is a way to undo a revert locally.

[the remark this answers is not in the document]
Comment (artkiver) [reply devel/got/Portfile]: > OK, ran this locally which hopefully undid that mistake?
> 
> ```
> git cherry-pick b280d80f2c389f52457cd9bbcedb098a5da46eec
> Auto-merging devel/got/Portfile
> [got0.92_uno c392269a321] got: update to 0.92
>  Date: Wed Aug 30 08:42:52 2023 +0000
>  1 file changed, 4 insertions(+), 4 deletions(-)
> ```
> 
> I'll find out when I am on a better network and can push the local changes I guess.

Well, that didn't seem to do the trick, I tried git reset to that commit as well to no avail; so pulling the other commits and then manually getting rid of them seemed to get my local branch to a state where I could push again.

There is maybe a better way to do that for git wizards, which I am not.

**Comment by artkiver** [comment]
That last revert attempt seems to have messed things up even more. ;-/

**Comment by artkiver** [comment]
In lieu of being able to clean up the commits due to my own ineptitude, superseded by https://github.com/macports/macports-ports/pull/20280
